libs/libc/stdio/lib_fmemopen.c

```diff
@@ -0,0 +1,291 @@
+    }
+
+  memcpy(buf, fmemopen_cookie->buf + fmemopen_cookie->pos, size);
+
+  fmemopen_cookie->pos += size;
+  return size;
+}
+
+/****************************************************************************
+ * Name: fmemopen_write
+ ****************************************************************************/
+
+static ssize_t fmemopen_write(FAR void *c, FAR const char *buf, size_t size)
+{
+  FAR struct fmemopen_cookie_s *fmemopen_cookie =
+    (FAR struct fmemopen_cookie_s *)c;
+  if (size + fmemopen_cookie->pos > fmemopen_cookie->size)
+    {
+      return -ENOSPC;
+    }
+
+  memcpy(fmemopen_cookie->buf + fmemopen_cookie->pos, buf, size);
+
+  fmemopen_cookie->pos += size;
+  if (fmemopen_cookie->pos > fmemopen_cookie->end)
+    {
+      fmemopen_cookie->end = fmemopen_cookie->pos;
+    }
+
+  /* POSIX states that NULL byte shall be written at the current position
+   * or end of the buffer.
+   */
+
+  if (fmemopen_cookie->pos < fmemopen_cookie->size &&
+      fmemopen_cookie->buf[fmemopen_cookie->pos - 1] != '\0')
+    {
+      fmemopen_cookie->buf[fmemopen_cookie->pos] = '\0';
+    }
+
+  return size;
+}
+
+/****************************************************************************
+ * Name: fmemopen_seek
+ ****************************************************************************/
+
+static off_t fmemopen_seek(FAR void *c, FAR off_t *offset, int whence)
+{
+  FAR struct fmemopen_cookie_s *fmemopen_cookie =
+    (FAR struct fmemopen_cookie_s *)c;
+  off_t new_offset;
+
+  switch (whence)
+    {
+      case SEEK_SET:
+        new_offset = *offset;
+        break;
+      case SEEK_END:
+        new_offset = fmemopen_cookie->end + *offset;
+        break;
+      case SEEK_CUR:
+        new_offset = fmemopen_cookie->pos + *offset;
+        break;
+      default:
+        return ENOTSUP;
+    }
+
+  /* Seek to negative value or value larger than maximum size shall fail. */
+
+  if (new_offset < 0 || (new_offset > fmemopen_cookie->end))
+    {
+      return EINVAL;
+    }
+
+  fmemopen_cookie->pos = new_offset;
+  *offset = new_offset;
+  return new_offset;
+}
+
+/****************************************************************************
+ * Name: fmemopen_close
+ ****************************************************************************/
+
+static int fmemopen_close(FAR void *c)
+{
+  FAR struct fmemopen_cookie_s *fmemopen_cookie =
+    (FAR struct fmemopen_cookie_s *)c;
+  if (fmemopen_cookie->custom)
+    {
+      lib_free(fmemopen_cookie->buf);
+    }
+
+  lib_free(fmemopen_cookie);
+  return 0;
+}
+
+/****************************************************************************
+ * Public Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: fmemopen
+ ****************************************************************************/
+
+FAR FILE *fmemopen(FAR void *buf, size_t size, FAR const char *mode)
```

Comment:
should we implement a similar function too:
https://man7.org/linux/man-pages/man3/open_memstream.3.html

Reply:
Yes, it is in my to do list for next (or the one after next) week. But I plan it to implement it to separate file (`lib_open_memstream` probably) together with `open_wmemstream`.